File: app/src/main/java/ca/appvelopers/mcgillmobile/model/Place.java
/*
 * Copyright 2014-2016 Appvelopers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.appvelopers.mcgillmobile.model;

import com.google.android.gms.maps.model.LatLng;

import java.io.Serializable;
import java.util.List;

/**
 * A place on the campus map
 * @author Julien Guerinet
 * @since 1.0.0
 */
public class Place implements Serializable {
    private static final long serialVersionUID = 1L;
    /**
     * The place name
     */
    protected String name;
    /**
     * The place categories
     */
    protected List<Integer> categories;
    /**
     * The address of this place
     */
    protected String address;
    /**
     * The latitude coordinate of this place
     */
    protected double latitude;
    /**
     * The longitude coordinate of this place
     */
    protected double longitude;

    /**
     * Default Moshi Constructor
     */
    protected Place() {}

    /* GETTERS */

    /**
     * @return The name of this place
     */
    public String getName() {
        return name;
    }

    /**
     * @return The address of this place
     */
    public String getAddress() {
        return address;
    }

    /**
     * @return The place coordinates
     */
    public LatLng getCoordinates() {
        return new LatLng(latitude, longitude);
    }

    /**
     * Checks if this place is of the given type
     *
     * @param type The type
     * @return True if it is part of the type, false otherwise
     */
    public boolean isOfType(PlaceType type) {
        //TODO
        return false;
//        return types.contains(type);
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof Place)) {
            return false;
        }

        Place place = (Place) object;
        return name.equalsIgnoreCase(place.getName()) &&
                address.equalsIgnoreCase(place.getAddress()) &&
                getCoordinates().equals(place.getCoordinates());
    }
}

File: app/src/main/java/model/place/Place.kt
/*
 * Copyright 2014-2019 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.model.place

import com.google.firebase.firestore.GeoPoint
import com.guerinet.mymartlet.util.Constants
import com.guerinet.mymartlet.util.extensions.get
import com.guerinet.mymartlet.util.firestore
import com.guerinet.suitcase.coroutines.ioDispatcher
import kotlinx.coroutines.withContext
import timber.log.Timber

/**
 * A place on the campus map
 * @author Julien Guerinet
 * @since 1.0.0
 */
data class Place(
    val id: Int,
    val name: String,
    val categories: List<Int>,
    val address: String,
    private val courseName: String,
    val coordinates: GeoPoint
) {

    /** Name of the place when listed under a course location, an empty String if equivalent to the [name]  */
    val coursePlaceName: String
        // If there is no override, simply use the name
        get() = courseName.takeIf { it.isNotEmpty() } ?: name

    companion object {

        /**
         * Loads the places from the Firestore
         */
        suspend fun loadPlaces(): List<Place> = withContext(ioDispatcher) {
            firestore.get(Constants.Firebase.PLACES) {
                val id = it.id.toInt()
                val name = it["name"] as? String
                @Suppress("UNCHECKED_CAST")
                val categories = it["categories"] as? List<Long>
                val address = it["address"] as? String
                val courseName = it["courseName"] as? String ?: ""
                val coordinates = it["coordinates"] as? GeoPoint

                if (name != null && categories != null && address != null && coordinates != null) {
                    // Note: not checking courseName here as it is an optional field, and if missing can be
                    //  safely cast to an empty String
                    Place(id, name, categories.map { category -> category.toInt() }, address, courseName, coordinates)
                } else {
                    Timber.tag("LoadPlaces").e(
                        Exception(
                            "Error parsing Place with id $id. name: $name, " +
                                "categories: $categories, address: $address, coordinates: $coordinates"
                        )
                    )
                    null
                }
            }
        }
    }
}
